### ISSUE TITLE
Insertion

### ISSUE THREAD

**Issue by fredefox** [MUSTE-Project/MULLE#2]
> We need a way for users to insert new nodes into adjunction trees.
> This may be achieved by simply having the back alternate between
> suggesting replacements for syntactically close nodes.

&mdash; paraphrasing

As far as I can tell the above proposal will not actually allow for
insertions.  At least not on its own.  When I click a word it only
allows replacing that part of the sentence.  So if we want to
implement it like this, then we must change how the core logic deals
with creating adjunction trees.

Furthermore, the front-end logic determines how far up the tree we are
querying by seeing if the text is we are clicking is stroked through.
We would need to change this.

These two considerations taken into account makes me think that
perhaps we should come up with another idea on how to implement this.

**Comment by fredefox** [comment]
Any feedback on this?

Do you think I should tackle this before or after #6 since this issue will also touch the front-end logic. I could probably implement this with minimal changes to the front-end logic, but of course it will be a little bit of double work.

**Comment by heatherleaf** [comment]
I think the current javascript has support for insertion, it’s just not used right now. Don’t remember how it works though... probably needs some changes to the Ajax api

**Comment by heatherleaf** [comment]
Here’s an idea how to implement insertion:

Assume we’re building the suggestions for subtree s somewhere inside the tree. For simplicity, let s cover the words w_j...w_k in the linearisation (w_1...w_n). Every word w_j...w_k are introduced by some node s_j, ..., s_k (all s_i are in subtree s, and s_i and s_i’ can be the same node (but doesn’t have to))

Now, collect all replacement subtrees for s (Prune.replaceTrees) and look through them. Given a replacement r, let its linearisation cover be u_p...u_q, and their corresponding nodes be r_p, ..., r_q: 

IF all cover nodes of s (s_j, ..., s_k) are included in r_p, ..., r_q; 
AND the linearisation order between s_j, ..., s_k are kept the same as in w_j...w_k; 
AND there are some additional new cover nodes r_i, ..., r_i’ which are not in s_j, ..., s_k; 
THEN r should be an insertion (at the corresponding positions of r_i, ..., r_i’);
OTHERWISE r is a normal replacement

I don’t know how informative that was. Shout if you don’t get it.

**Comment by fredefox** [comment]
Let me see if I get this, on a higher level: The idea is to change the type of the value returned by `replaceTrees` from

    Map Path [(Int, TTree)]

To something like (other options exist, e.g. two maps, one for the internal nodes and one for all the spaces in between)

    Map Path [(Int, Bool, TTree)]

Where the `Bool` indicates if the replacement is to be considered an insertion.

Then in the UI when toggling through the different suggestions for a node we only show the ones that are not insertions, and when clicking on a space between two nodes we toggle through the appropriate nodes that correspond to insertion at the given point.

**Comment by fredefox** [comment]
Let me see if I get this, on a higher level: The idea is to change the type of the value returned by `replaceTrees` from

    Map Path [(Int, TTree)]

To something like (other options exist, e.g. two maps, one for the internal nodes and one for all the spaces in between)

    Map Path [(Int, Bool, TTree)]

Where the `Bool` indicates if the replacement is to be considered an insertion.

Then in the UI when toggling through the different suggestions for a node we only show the ones that are not insertions, and when clicking on a space between two nodes we toggle through the appropriate nodes that correspond to insertion at the given point.

**Comment by heatherleaf** [comment]
Yes, something like that. Note that N(insertion points) = N(words) + 1. I don’t know which representation is the best. 

Also note that an insertion can come before or after (or even in between) the words in the phrase.

**Comment by fredefox** [comment]
FYI: I'm starting work on this now.

**Comment by fredefox** [comment]
Being developed here: [`wip/2-insertion`](https://github.com/MUSTE-Project/MULLE/tree/wip/2-insertion). I've changed the representation so far. Now I just need to

- [ ] implement your algorithm
- [ ] have the UI handle it as well

**Comment by fredefox** [comment]
My head is spinning a bit now and I have gotten a bit "snow blind". Think I need to step out for a few seconds.

I've copied your spec to the location where I think it makes most sense to implement [`costTree`](https://github.com/MUSTE-Project/MULLE/blob/2840713fa7e79ebabc41bc7e6ebfeccca83f5d9e/muste/src/Muste/Menu.hs#L77)

My current understanding is that we just need to compare the parameter `s` (should be `r`) to an additional parameter `s` that we can get from the caller `getPrunedSuggestions` above. The comparison is c.f. your spec.

**Comment by fredefox** [comment]
I have implemented your algorithm sans the notion of a "cover node":

https://github.com/MUSTE-Project/MULLE/blob/af1740d4701f9ae719cb52a38ff8283b5a7f2c51/muste/src/Muste/Menu.hs#L120-L122

Can you elaborate?

**Comment by heatherleaf** [comment]
Hmm... it's starting to get too hot now, but I'll have a try:

The "cover" of a subtree *t* (of the main tree *T*) are the `LinToken`s that are introduced by *t*, i.e. a list of `LinToken`. The "cover node" of a `LinToken` is the specific tree node in *t* that introduced the token, i.e., the node referred to by `_ltpath`.

Does that clarify better?

**Comment by heatherleaf** [comment]
Here’s a more complete attempt to define cover nodes. 

The subtree *t* (of *T*) is referenced by path *p*. The linearisation of *T* is

    lin_T = [(w_1,p_1), (w_2,p_2) ..., (w_n,p_n)]

The cover of *t* is

    cover_t = [ tok’ | tok’@(_,p’) <- lin_T, isPrefix(p,p’) ]

The cover nodes are `[ p’ | (_,p’) <- cover_t ]`

**Comment by heatherleaf** [comment]
So, assume *r* is the new suggested subtree replacement for *t* (and *R* is the full tree where *t* is replaced for *r*). 

If `coverNodes_t` is a sublist of `coverNodes_r` (all nodes in `coverNodes_t` occur in `coverNodes_r` and in the same order), then *r* is an insertion. 

(By sublist I mean that the list doesn’t have to be continuous, but the order must be the same. Eg, “135” is a sublist of “12345”)

**Comment by heatherleaf** [comment]
One complication to handle is this: the path to a tree node can change during a replacement. Eg, assume this tree:

    (a b (c d))

And insert `(f g h * k)` at the `c` node to get this new tree:

    (a b (f g h (c d) k))

Now the paths for the `c` and `d` nodes have changed (from [2] resp [21] to [23] resp [231]).

**Comment by fredefox** [comment]
At your previous comment: This is why I think it might make sense to change the representation of `Linearization`. I have two initial ideas. Swap it out with a non-functional data-structure where we have `IORef`s in stead of of `Path`s. With some care we would be able to make a pure interface to this using `unsafePerformIO`. The downside of course is that we still need to handling converting to/from JSON. Converting *to* JSON shouldn't be so difficult, it's probably more the other way around.

Otherwise there is also mr. Kmett's [`graphs`](http://hackage.haskell.org/package/graphs) (probably implemented in a similar style as the above suggestion). I'm not sure that if we proceed this way, that we will be able to keep the same internal representation for `TTree`. I could also make this type abstract and instance of the mono-traversable as I have done for so many of the other types - the challenge is that the shape of `TTree` doesn't quite fit so elegantly what Snoyberg envisioned when he designed `mono-traversable` -- `TTree`s in a sense contains `Either (String, FunType) Category`.

**Comment by heatherleaf** [comment]
Yes, I have also thought about using pointers (`IORef`). If you can find a way to use `IORef` and still have the application safe, stable and understandable, I'm all for it.

If we want to keep paths as references, then it should be possible calculate a bijection between paths from the pruned subtrees of a `SimTree`.

**Comment by fredefox** [comment]
It would be good with a test case I could add for this. In particular I had in mind creating a test-case for

    isInsertion :: Context -> Path -> TTree -> TTree -> Bool

I just mention for future reference that we must a grammar (hence `Context`) because the linearization is part of the algorithm.

**Comment by fredefox** [comment]
@daherb , last Thursday we talked about changing how we determine if something is a linearization. I'm afraid the notes I took were not detailed enough for me to start implementing this. Can you refresh my memory? We wanted to determine this by looking at the "inner nodes" corresponding to the linearization.

I'm guessing that what we need to change is:

    coverNodes :: Context -> Path -> TTree -> [Path]

which is used in `isInsertion` to create a comparison between trees by checking that the cover nodes from one is a sub-list of the other.

**Comment by heatherleaf** [comment]
I think I remember @daherb 's suggestion, it was a simplification of my previous idea. First, we don't have to check any `Path` in `isInsertion`, and it can return at which position(s) it will be inserted:

    isInsertion :: Context -> TTree -> TTree -> [Int]

If it returns `[]`, it means it's not an insertion. If it returns `[2]`, it means that it's an insertion directly before token 2. If it returns `[2,5]` it's an insertion in two positions (which could happen for uncommon discontinuities).

Here's an attempt. I haven't tested, not even tried to compile, but I hope that you understand what I'm trying to do:

    type CoverNode = (Int, String, Path)

    coverNodes :: Context -> TTree -> [CoverNode]
    coverNodes cxt tree = do
        let Linearization tokens = linearizeTree cxt tree
        (n, LinToken path _ _) <- zip [0..] tokens
        let Just (TNode fun _ _) = Tree.selectNode tree path
        return (n, fun, path)

    isInsertion :: Context -> TTree -> TTree -> [Int]
    isInsertion cxt src trg = [ n | (n, _) <- inserted ]
        where
            srcnodes = coverNodes cxt src
            trgnodes = coverNodes cxt trg 
            inserted = fromMaybe [] (getInsertedNodes srcnodes trgnodes)

            getInsertedNodes :: [CoverNode] -> [CoverNode] -> Maybe [(Int, CoverNode))]
            getInsertedNodes [] [] = Just []
            getInsertedNodes _ [] = Nothing
            getInsertedNodes srcs@((_,f,_):_) ((_,g,p):trgs@((_,g',p'):_)) 
                | f == g && g == g' && p == p'  = getInsertedNodes srcs trgs
            getInsertedNodes ((_,f,_):srcs) ((_,g,_):trgs)
                | f == g  = getInsertedNodes srcs trgs
            getInsertedNodes srcs (node:trgs)
                = fmap ((getInsertionPosition srcs,node) : ) (getInsertedNodes srcs trgs)

            getInsertionPosition ((n,_,_):_) = n
            getInsertionPosition [] = length srcnodes

**Comment by fredefox** [comment]
I've added your implementation to the code-base with a few modifications. I'll change it so that we also use it now.

**Comment by fredefox** [comment]
At your most recent comment: I don't think efficiency is a concern for 'Selection', so I'm not sure we need the last representation -- unless I don't fully understand the purpose of it.

The idea of incorporating invariants is easier since I've made `Selection` abstract (i.e. i'm not exporting the constructor) so if we can easily make some kind of smart constructor like this. Currently we just have `fromList :: [Int] -> Selection` which makes no such checks, but it could conceivable be changed to `fromList :: [Int] -> Maybe Selection` to enforce some kind of invariant (they're called "smart" constructors).

My gut feeling is to go for the first representation since it's by far the simplest and seems to be able to support what we want. We can then define smart constructors like:

    words :: [Int] -> Selection
    spaces :: [Int] -> Selection
    wordAndSpaces :: [Int] -> [Int] -> Selection

and use the correct ones where appropriate.

**Comment by fredefox** [comment]
I'm assuming you meant:

> But insertion isn't [sic] between tokens

**Comment by fredefox** [comment]
Running

    pretty $ concatMap snd $ Mono.mapToList $ getMenuFromStringRep theCtxt "fienden besegrar Afrika"

Returns [this](https://pastebin.com/raw/AfSSUdKe) with the `Pretty` instance I made for `CostTree`s. (it prints `INS: ` before the sentence if it's an insertion.)

**Comment by fredefox** [comment]
`NewFancyMenu` does not record which nodes are considered insertions. Can you be a bit more specific about what you mean?

**Comment by fredefox** [comment]
I suppose you're referring to the highlighting mechanism in the cli. Let's have a closer look. Say we have a menu item like:

    flickan [ stänger dålig ] på datorn

If we insert clickable spaces here, what should it look like? I can think of a range of valid ideas:

    _ flickan _ [ _ stänger _ dålig _ ] _ på _ datorn _
    _ flickan _ [   stänger _ dålig _ ]   på _ datorn _
    _ flickan   [ _ stänger _ dålig   ] _ på _ datorn _
    _ flickan _ [   stänger _ dålig   ] _ på _ datorn _
    _ flickan   [ _ stänger _ dålig _ ]   på _ datorn _
    _ flickan   [   stänger _ dålig   ]   på _ datorn _

**Comment by fredefox** [comment]
I have a first draft in the commit referenced above. Please have a look at that. What it does is simply look for all the "empty" selections and select the ones corresponding to the clicked space.

**Comment by heatherleaf** [comment]
A selection {k-k} is an insertion. E.g., if I run the CLI with "en flicka läser", I get the following menu:

```
{1-1}: en [ ] flicka läser
en [ blå ] flicka läser
en [ dålig ] flicka läser
en [ god ] flicka läser
en [ kall ] flicka läser
en [ stor ] flicka läser
```
{1-1} here means that the menu should be inserted between "en" and "flicka".

Another menu should be in two places, both if you click before "en", and if you click on "läser":
```
{0-0,2-3}: [ ] en flicka [ läser ]
[ idag läser ] en flicka [ ]
[ nu läser ] en flicka [ ]
[ redan läser ] en flicka [ ]
```

**Comment by heatherleaf** [comment]
Looks good! I don't see any problems 

I note that you can handle two different menus at the same insertion point. E.g., one menu with selection {2-2} and another with selection {0-1,2-2}:

```
{0-1,2-2}: [ kungen ] älskar [ ] böcker
[ idag ] älskar [ kungen ] böcker
[ nu ] älskar [ kungen ] böcker

{2-2}: kungen älskar [ ] böcker
kungen älskar [ blåa ] böcker
kungen älskar [ dåliga ] böcker
```

It nicely alternates between the two when I click the insertion point.

**Comment by fredefox** [comment]
Yay!
